mobilecoind/src/transaction_memo.rs

```diff
@@ -0,0 +1,335 @@
+    /// subaddress index to generate the sender memo credential from
+    Rth {
+        /// Optional subaddress index to generate the sender memo credential
+        /// from.
+        subaddress_index: Option<u64>,
+    },
+
+    RthWithPaymentIntentId {
+        /// Optional subaddress index to generate the sender memo credential
+        /// from.
+        subaddress_index: Option<u64>,
+
+        /// The payment intent id to include in the memo.
+        payment_intent_id: u64,
+    },
+
+    RthWithPaymentRequestId {
+        /// Optional subaddress index to generate the sender memo credential
+        /// from.
+        subaddress_index: Option<u64>,
+
+        /// The payment request id to include in the memo.
+        payment_request_id: u64,
+    },
+
+    /// Empty Transaction Memo.
+    Empty,
+
+    /// Burn Redemption memo.
+    BurnRedemption([u8; BurnRedemptionMemo::MEMO_DATA_LEN]),
+}
+
+impl TransactionMemo {
+    pub fn memo_builder(&self, account_key: &AccountKey) -> Box<dyn MemoBuilder + Send + Sync> {
+        match self {
+            Self::Empty => Box::<EmptyMemoBuilder>::default(),
+            Self::Rth { subaddress_index } => {
+                let memo_builder = generate_rth_memo_builder(subaddress_index, account_key);
+                Box::new(memo_builder)
+            }
+            Self::RthWithPaymentIntentId {
+                subaddress_index,
+                payment_intent_id,
+            } => {
+                let mut memo_builder = generate_rth_memo_builder(subaddress_index, account_key);
+                memo_builder.set_payment_intent_id(*payment_intent_id);
+                Box::new(memo_builder)
+            }
+            Self::RthWithPaymentRequestId {
+                subaddress_index,
+                payment_request_id,
+            } => {
+                let mut memo_builder = generate_rth_memo_builder(subaddress_index, account_key);
+                memo_builder.set_payment_request_id(*payment_request_id);
+                Box::new(memo_builder)
+            }
+            Self::BurnRedemption(memo_data) => {
+                let mut memo_builder = BurnRedemptionMemoBuilder::new(*memo_data);
+                memo_builder.enable_destination_memo();
+                Box::new(memo_builder)
+            }
+        }
+    }
+}
+
+fn generate_rth_memo_builder(
+    subaddress_index: &Option<u64>,
+    account_key: &AccountKey,
+) -> RTHMemoBuilder {
+    let mut memo_builder = RTHMemoBuilder::default();
+    let sender_memo_credential = match subaddress_index {
+        Some(subaddress_index) => SenderMemoCredential::new_from_address_and_spend_private_key(
+            &account_key.subaddress(*subaddress_index),
+            account_key.subaddress_spend_private(*subaddress_index),
+        ),
+        None => SenderMemoCredential::from(account_key),
+    };
+    memo_builder.set_sender_credential(sender_memo_credential);
+    memo_builder.enable_destination_memo();
```

Comment:
we might want this to be optional also, the previous behavior didn't send any memos I believe. even though sending these memos is fairly harmless, someone might want to be able to get the old behavior for some reason

Reply:
If you don't want to send memos you can set the TransactionMemo to empty (https://github.com/mobilecoinfoundation/mobilecoin/pull/3945/files/cacecd2c995ff52cec4bbdf00d630bc9e0300e5c#diff-de0af4379aad1f14d9af009735bb3d91767a7d5e65d33b09c8d45f484a5c2324R369)

Reply:
i see, that seems fine